synapse/appservice/api.py

```diff
@@ -132,11 +133,16 @@ async def query_user(self, service: "ApplicationService", user_id: str) -> bool:
         assert service.hs_token is not None
 
         try:
-            response = await self.get_json(
-                f"{service.url}{APP_SERVICE_PREFIX}/users/{urllib.parse.quote(user_id)}",
-                {"access_token": service.hs_token},
-                headers={"Authorization": [f"Bearer {service.hs_token}"]},
-            )
+            if self.config.use_appservice_legacy_authorization:
+                response = await self.get_json(
+                    f"{service.url}{APP_SERVICE_PREFIX}/users/{urllib.parse.quote(user_id)}",
+                    {"access_token": service.hs_token},
+                )
+            else:
+                response = await self.get_json(
+                    f"{service.url}{APP_SERVICE_PREFIX}/users/{urllib.parse.quote(user_id)}",
+                    headers={"Authorization": [f"Bearer {service.hs_token}"]},
+                )
```

Comment:
I don't think this is quite what I expected, I expected:

1. When `use_appservice_legacy_authorization=false` only the authorization header would be set.
2. When `use_appservice_legacy_authorization=true` both the authorizaiton header and query parameter would be set.

I.e. option 2 is essentially "what we do today". We don't want to send *only* the query parameter because that gives people no way to upgrade.

I think we don't want to copy the whole request twice though. I'd suggest something like:

```suggestion
            args = None
            if self.config.use_appservice_legacy_authorization:
                args = {"access_token": service.hs_token}
            response = await self.get_json(
                f"{service.url}{APP_SERVICE_PREFIX}/users/{urllib.parse.quote(user_id)}",
                args,
                headers={"Authorization": [f"Bearer {service.hs_token}"]},
            )
```